include/envoy/router/route_config_provider_manager.h

```diff
@@ -27,43 +27,41 @@ class RouteConfigProviderManager {
 
   /**
-   * Get a RouteConfigProviderSharedPtr for a route from RDS. Ownership of the RouteConfigProvider
-   * is shared by all the HttpConnectionManagers who own a RouteConfigProviderSharedPtr. The
-   * RouteConfigProviderManager holds weak_ptrs to the RouteConfigProviders. Clean up of the weak
-   * ptrs happen from the destructor of the RouteConfigProvider. This function creates a
-   * RouteConfigProvider if there isn't one with the same (route_config_name, cluster) already.
-   * Otherwise, it returns a RouteConfigProviderSharedPtr created from the manager held weak_ptr.
+   * Get a RouteConfigProviderPtr for a route from RDS. Ownership of the RouteConfigProvider is the
+   * HttpConnectionManagers who calls this function. The RouteConfigProviderManager holds raw
+   * pointers to the RouteConfigProviders. Clean up of the pointers happen from the destructor of
+   * the RouteConfigProvider. This method creates a RouteConfigProvider which may share the
+   * underlying RDS subscription with the same (route_config_name, cluster).
    * @param rds supplies the proto configuration of an RDS-configured RouteConfigProvider.
    * @param factory_context is the context to use for the route config provider.
    * @param stat_prefix supplies the stat_prefix to use for the provider stats.
    */
-  virtual RouteConfigProviderSharedPtr getRdsRouteConfigProvider(
+  virtual RouteConfigProviderPtr createRdsRouteConfigProvider(
       const envoy::config::filter::network::http_connection_manager::v2::Rds& rds,
       Server::Configuration::FactoryContext& factory_context, const std::string& stat_prefix) PURE;
 
   /**
    * Get a RouteConfigSharedPtr for a statically defined route. Ownership is as described for
-   * getRdsRouteConfigProvider above. Unlike getRdsRouteConfigProvider(), this method always creates
-   * a new RouteConfigProvider.
+   * getRdsRouteConfigProvider above. This method always create a new RouteConfigProvider.
    * @param route_config supplies the RouteConfiguration for this route
    * @param runtime supplies the runtime loader.
    * @param cm supplies the ClusterManager.
    */
-  virtual RouteConfigProviderSharedPtr
-  getStaticRouteConfigProvider(const envoy::api::v2::RouteConfiguration& route_config,
-                               Server::Configuration::FactoryContext& factory_context) PURE;
+  virtual RouteConfigProviderPtr
+  createStaticRouteConfigProvider(const envoy::api::v2::RouteConfiguration& route_config,
+                                  Server::Configuration::FactoryContext& factory_context) PURE;
 
   /**
-   * @return std::vector<Router::RouteConfigProviderSharedPtr> a list of all the
+   * @return std::vector<Router::RouteConfigProvider*> a list of all the
    * dynamic (RDS) RouteConfigProviders currently loaded.
    */
-  virtual std::vector<RouteConfigProviderSharedPtr> getRdsRouteConfigProviders() PURE;
+  virtual std::vector<RouteConfigProvider*> getRdsRouteConfigProviders() PURE;
```

Comment:
nit: Can we potentially return std::reference_wrappers here and below? Since these can never be null it would be a bit cleaner.

Reply:
It is better to return list of RdsSubscription which is what the manager is holding.    This function is only used for dumping RDS config.  RdsSubscription is good enough

Reply:
RdsSubscription is not implementing the RouteConfigProvider interface and there is no corresponding interface.

Do we really need these two methods (get{Rds,Static}RouteConfigProviders) in RouteConfigProviderManager interface? It is only used in rds_impl_test and RouteConfigProviderManagerImpl itself to dump config. cc @jsedgwick @htuch

Reply:
I don't recall the entire history here. It's possible the methods are not needed anymore with iterative changes. Feel free to remove them if it's possible.

Reply:
+1, remove any methods from the interface that don't have external callers.